lib/codec_choose.c

```diff
@@ -195,70 +220,179 @@ codec_choose_x86 (struct codec *codec)
+	//    (allowing saving YMM registers on context switch)
+	// 2) CPUID indicates support for AVX
+	// 3) XGETBV indicates the AVX registers will be saved and restored on
+	//    context switch
+	//
+	// Note that XGETBV is only available on 686 or later CPUs, so the
+	// instruction needs to be conditionally run.
+	if (max_level >= 1) {
+		__cpuid_count(1, 0, eax, ebx, ecx, edx);
+		if (ecx & bit_XSAVE_XRSTORE) {
+			uint64_t xcr_mask;
+			xcr_mask = _xgetbv(_XCR_XFEATURE_ENABLED_MASK);
+			if (xcr_mask & _XCR_XMM_AND_YMM_STATE_ENABLED_BY_OS) {
 				__cpuid_count(1, 0, eax, ebx, ecx, edx);
 				if (ecx & bit_AVX) {
-					codec->enc = base64_stream_encode_avx;
-					codec->dec = base64_stream_decode_avx;
 					return true;
 				}
-				#endif
 			}
 		}
 	}
-	#endif
+	return false;
+}
+#else
+static bool avx_supported(void) {return false;}
+#endif
+
+#if HAVE_SSE42
+static bool sse42_supported(void)
+{
+	unsigned int eax, ebx = 0, ecx = 0, edx;
+	unsigned int max_level;
+
+#ifdef _MSC_VER
+	int info[4];
+	__cpuidex(info, 0, 0);
+	max_level = info[0];
+#else
+	max_level = __get_cpuid_max(0, NULL);
+#endif
 
-	#if HAVE_SSE42
 	// Check for SSE42 support:
 	if (max_level >= 1) {
 		__cpuid(1, eax, ebx, ecx, edx);
 		if (ecx & bit_SSE42) {
-			codec->enc = base64_stream_encode_sse42;
-			codec->dec = base64_stream_decode_sse42;
 			return true;
 		}
 	}
-	#endif
+	return false;
+}
+#else
+static bool sse42_supported(void) {return false;}
+#endif
+
+#if HAVE_SSE41
+static bool sse41_supported(void)
+{
+	unsigned int eax, ebx = 0, ecx = 0, edx;
+	unsigned int max_level;
+
+#ifdef _MSC_VER
+	int info[4];
+	__cpuidex(info, 0, 0);
+	max_level = info[0];
+#else
+	max_level = __get_cpuid_max(0, NULL);
+#endif
 
-	#if HAVE_SSE41
 	// Check for SSE41 support:
 	if (max_level >= 1) {
 		__cpuid(1, eax, ebx, ecx, edx);
 		if (ecx & bit_SSE41) {
-			codec->enc = base64_stream_encode_sse41;
-			codec->dec = base64_stream_decode_sse41;
 			return true;
 		}
 	}
-	#endif
 
-	#if HAVE_SSSE3
+	return false;
+}
+#else
+static bool sse41_supported(void) {return false;}
+#endif
+
+#if HAVE_SSSE3
+static bool ssse3_supported(void)
+{
+	unsigned int eax, ebx = 0, ecx = 0, edx;
+	unsigned int max_level;
+
+#ifdef _MSC_VER
+	int info[4];
+	__cpuidex(info, 0, 0);
+	max_level = info[0];
+#else
+	max_level = __get_cpuid_max(0, NULL);
+#endif
+
 	// Check for SSSE3 support:
 	if (max_level >= 1) {
 		__cpuid(1, eax, ebx, ecx, edx);
 		if (ecx & bit_SSSE3) {
-			codec->enc = base64_stream_encode_ssse3;
-			codec->dec = base64_stream_decode_ssse3;
 			return true;
 		}
 	}
-	#endif
 
+	return false;
+}
 #else
-	(void)codec;
+static bool ssse3_supported(void) {return false;}
 #endif
 
-	return false;
+static bool
+codec_choose_x86 (struct codec *codec)
+{
+    if(avx2_supported()) {
+	    codec->enc = base64_stream_encode_avx2;
+	    codec->dec = base64_stream_decode_avx2;
+	    return true;
+    };
+    if( ssse3_supported()) {
```

Comment:
Oops, copy/paste error, should be avx_supported()

Reply:
Also, the code style dictates a space after `if`:

```c
if (avx_supported()) {
```

But I can fix that in the rebase.